docs: plugin tests & minor improvement (#2067)
sopel-irc/sopel: docs/source/plugin/test.rst

.. _plugin-test:

===============
Automated tests
===============

Testing a plugin manually can become tedious and it is an activity prone to
human mistakes. To automate testing, Sopel provides a

> :mod:`pytest plugin <sopel.tests.pytest_plugin>` , with a set of testing tools

Author: Oh no, the after-space is just a mistake.

        return botfactory.preloaded(settings, ['myplugin'])

    @pytest.fixture
    def irc(bot, ircfactory):
        return ircfactory(bot)

    @pytest.fixture
    def user(userfactory):
        return userfactory('MyNick')

    @pytest.fixture
    def owner(userfactory):
        return userfactory('OwnerNick')

    def test_my_command(bot, irc, user):
        irc.pm(user, '.mycommand arg')

        assert bot.backend.message_sent == rawlist(
            'PRIVMSG MyNick :Command answer to a regular user.'
        )

    def test_my_command_owner(bot, irc, owner):
        irc.pm(owner, '.mycommand arg')

        assert bot.backend.message_sent == rawlist(
            'PRIVMSG MyNick :Command answer to my owner.'
        )

Test setups
===========

Before you can actually test a rule or a command, you will need some setups:

* a test configuration
* a test bot
* a test server and test users
* or a test trigger

For that, Sopel provides factories through pytest fixtures. In the above
example, these factories are used to create custom pytest fixtures.

.. seealso::

    Sopel's tests use a lot of fixtures and custom fixtures. Check the

>     `pytest fixture documentation`__ for more information about fixtures.

Author: Custom fixtures, as opposed to fixtures given by pytest's plugins, I'd say.

>     `pytest fixture documentation`__ for more information about fixtures.

Author: It should be better now, with way less repetition of the word "fixture".

        tmpconfig.define_section('myplugin', MyPluginSection)

And since we are using pytest, you can create your own local fixture for that::

    @pytest.fixture
    def tmpconfig(configfactory):
        return configfactory(TEST_NAME, TMP_CONFIG)

    def test_my_command(tmpconfig):
        tmpconfig.define_section('myplugin', MyPluginSection)

If all you need is a test configuration, you could put the section definition
within your fixture. However, there are other options available to you when you
use a test bot.

Test bot
--------

Now that you have a test configuration available through your custom
``tmpconfig`` fixture, you may want a test bot. Use the
:func:`~sopel.tests.pytest_plugin.botfactory` fixture for that::

    def test_my_command(tmpconfig, botfactory):
        bot = botfactory(tmpconfig)

However at this point, the bot doesn't know about your plugin, so it hasn't
run the setup phase. You can do that with the
:meth:`~sopel.tests.factories.BotFactory.preloaded` method::

    def test_my_command(tmpconfig, botfactory):
        bot = botfactory.preloaded(tmpconfig, ['myplugin'])
        assert bot.has_plugin('myplugin')  # should be True

> .. note::

Reviewer: The "must" tells me this should probably be:

```suggestion
.. important::
```

    by your ``setup`` plugin hook.

Of course, if you want to reuse the same test bot in all your tests, you can
create a fixture for that::

    @pytest.fixture
    def bot(configfactory, botfactory):
        settings = configfactory(TEST_NAME, TEST_CONFIG)
        return botfactory.preloaded(settings, ['myplugin'])

And then use it in your tests::

    def test_my_command(bot):
        assert bot.has_plugin('myplugin')  # should be True

The ``bot`` created by the factory is a regular instance of
:class:`sopel.bot.Sopel` with a
:class:`test IRC backend <sopel.tests.mocks.MockIRCBackend>` instead of a
regular one. This backend doesn't send anything over the network and instead
it registers everything into its ``message_sent`` list::

    from sopel.tests import rawlist

    def test_my_command(bot):
        bot.say('Hi!', '#channel')
        assert len(bot.backend.message_sent) == 1
        assert bot.backend.message_sent == rawlist(
            'PRIVMSG #channel :Hi!',
        )

.. seealso::

    For more information about the :func:`~sopel.tests.rawlist` function,
    see the `Checking the bot's output`_ section.

Test Server and test Users
--------------------------

> Alright, you have a properly setup test bot, now it's time for the last parts

Reviewer: ```suggestion
Now that you have a properly setup test bot, it's time for the last parts
```

This is the writer in me making things flow better. I didn't really like the comma splice. x)

Author: Right now, I like the writer in you.

of your test setup: a test server with test users. As usual, there are fixtures
to help you. The :func:`~sopel.tests.pytest_plugin.ircfactory` can be used to
create a test server, and the :func:`~sopel.tests.pytest_plugin.userfactory`
can create test users::

    @pytest.fixture
    def irc(bot, ircfactory):
        return ircfactory(bot)

    def test_my_command(bot, irc, userfactory):
        user = userfactory('MyNick')
        irc.pm(user, '.mycommand arg')

        assert bot.backend.message_sent == rawlist(
            'PRIVMSG MyNick :Command answer to a regular user.'
        )

    def test_my_command_owner(bot, irc, userfactory):
        owner = userfactory('OwnerNick')
        irc.pm(owner, '.mycommand arg')

        assert bot.backend.message_sent == rawlist(
            'PRIVMSG MyNick :Command answer to my owner.'
        )

> As usual, you can create custom fixtures for the test server and for your test

Reviewer: ```suggestion
As usual, you can create custom fixtures for the test server (as above) and for your test
```

Because it's been shown already, no need to repeat. But refer to the above example in case the user glosses over that part.

Author: See, that's why everyone needs an editor!

users, for example, one for a regular user, and one for the owner::

    @pytest.fixture
    def user(userfactory):
        return userfactory('MyNick')

    @pytest.fixture
    def owner(userfactory):
        return userfactory('OwnerNick')

Channel messages
................

A bot can join channels, and so does your test bot: by using the test server,
you can make the bot join a channel, or add new users to a channel while the

> bot is already in::

Reviewer: ```suggestion
bot is already in it::
```

:fr: turning :uk:? 😂

    def test_my_command(bot, irc, user, owner):
        # bot joins #channel with the owner in it
        irc.channel_join('#channel', users=[owner])

        # user joins #channel after
        irc.join(user, '#channel')

        # clear messages on join
        bot.backend.clear_message_sent()

        # user talks into a channel
        irc.say(user, '.mycommand arg')

        assert bot.backend.message_sent == rawlist(
            'PRIVMSG #channel :MyNick: my reply into a channel.'
        )

You can automate this setup within your fixture::

    @pytest.fixture
    def irc(bot, user, owner, ircfactory):
        test_server = ircfactory(bot)
        # auto-join channels
        test_server.channel_join('#channel', users=[owner, user])
        # clear messages on join
        bot.backend.clear_message_sent()
        return test_server

> And now you are all setup to test your plugin's commands and rules!

Reviewer: ```suggestion
And now you are all set up to test your plugin's commands and rules!
```

This compound word is never used correctly even by native speakers, I swear. 👀

Author: And I swear I have checked once or twice for another occurrence, but not that one. 😁

.. seealso::

    Sopel uses a lot of fixtures both from pytest and custom ones specificaly
    made for its test suite. Check the `pytest fixtures documentation`__ to

>     learn more about them as well as how to create yours.

Reviewer: ```suggestion
    learn more about them as well as how to create your own.
```

Now I'm just nitpicking.

Test Server and test Users
--------------------------

> Now that you have a properly setup test bot, it's time for the last parts

Reviewer: ```suggestion
Now that you have a test bot properly set up, it's time for the last parts
```

Yikes, reviewing on mobile last time got me.
